src: Allow `ga` to skip removed buffers

lenormf opened the pull request:
This commit prevents `ga` from returning a “no last buffer” error
when the previously displayed buffer was removed.

Since the jumps list keeps track of the order in which buffers were
displayed already, handling arbitrary `delete-buffer`s as well,
cycle through it to implement `ga` instead of storing a pointer.

Note that this commit doesn't take into account buffer flags that
might exclude some buffers from being cycled over by commands.

Fixes #1840

krobelus reviewed (commented):
Thanks, I've always wanted that behavior - I think this makes `ga` strictly better

> Note that this commit doesn't take into account buffer flags that
might exclude some buffers from being cycled over by commands.

I prefer that for the time being, I got used to use `ga` with the scratch buffer. But we can change that in future.

lenormf commented on src/context.cc:
```diff
@@ -197,16 +190,41 @@ void Context::change_buffer(Buffer& buffer)
+
+Buffer* Context::last_buffer() const
+{
+    if (not m_jump_list.size())
```
Thanks, fixed.

lenormf commented on src/context.cc:
```diff
@@ -197,16 +190,41 @@ void Context::change_buffer(Buffer& buffer)
+    if (not m_jump_list.size())
+        return nullptr;
+
+    const auto next_buffer = std::find_if(
```
I understand, I think the same way. But then I thought that `JumpList` *is* a vector, and the private data it uses (such as the `m_current` index) is only used by the extra methods that interact with the storage.

Since the intent behind this structure is not to override a container but only to add state to it, I think it's not that bad if it makes its use simpler.

@mawww just drop this commit if needed, I'm fine either way.

krobelus replied:
Yeah, probably both variants are fine. A compromise would be to expose a conversion operator (or a getter) returning `const Vector<Buffer*> &`, then this can stay roughly the same, and we can't accidentally modify the jumplist.

mawww commented on src/context.hh:
```diff
@@ -28,16 +28,14 @@ struct JumpList
     friend bool operator==(const JumpList& lhs, const JumpList& rhs)
     {
-        return lhs.m_jumps == rhs.m_jumps and lhs.m_current == rhs.m_current;
+        return lhs == rhs and lhs.m_current == rhs.m_current;
```
This looks like its going to be infinitely recursive, you'd need to cast to the base class or call vector::operator== explicitely, but either way I think it shows inheriting from std::vector might not be that nice here.

mawww commented on src/context.cc:
```diff
@@ -197,16 +191,51 @@ void Context::change_buffer(Buffer& buffer)
+
+Buffer* Context::last_buffer() const
+{
+    // FIXME: convert JumpList to a vector and use find_if() with reverse iterators
```
If you just expose a ConstArrayView from the jump list you can, you could even do `find_if(jump_list.list() | reverse(), <predicate>)` with the toolset exposed in ranges.hh

mawww reviewed (changes requested):
I like that this gets rid of the last buffer field and use the jump list instead, I would not want the jump list to inherit from Vector but you can just expose either begin/end iterators or a ConstArrayView to the data.

lenormf commented:
Rebased on master, dropped the commit that makes `JumpList` inherit its inner storage container, implement `JumpList::get_as_list()` to allow using regular algorithm functions.

mawww commented on src/context.cc:
```diff
@@ -197,16 +191,36 @@ void Context::change_buffer(Buffer& buffer)
+    };
+
+    auto next_buffer = std::find_if(jump_list.begin() + (m_jump_list.current_index()-1),
+                                    jump_list.end(), predicate);
```
`find_if(jump_list.subrange(m_jump_list.current_index() -1), predicate)` ?

mawww commented on src/context.cc:
```diff
@@ -197,16 +191,36 @@ void Context::change_buffer(Buffer& buffer)
+        return &next_buffer->buffer();
+
+    auto previous_buffer = std::find_if(jump_list.rbegin() + (jump_list.size() - m_jump_list.current_index()),
+                                        jump_list.rend(), predicate);
```
`find_if(jump_list.subrange(0, m_jump_list.current_index() | reverse(), predicate)` ?

mawww reviewed (changes requested):
Looks good, a couple nitpick on style.

lenormf commented:
Code factorised. Quite satisfied with how it looks now!
